daceml/autodiff/backward_pass_generator.py

```diff
@@ -0,0 +1,1270 @@
+            )
+
+        forward_subgraph = self._find_subgraph_to_differentiate()
+
+        # expand until there is nothing left to expand
+        while self._expand_nodes(forward_subgraph):
+            # Nodes have been expanded again on the expanded graph; recalculate the forward graph
+            forward_subgraph = self._find_subgraph_to_differentiate()
+
+        if self.apply_strict:
+            self.sdfg.apply_strict_transformations()
+            forward_subgraph = self._find_subgraph_to_differentiate()
+
+        # check that all edges are float
+        for edge, parent_subgraph in forward_subgraph.all_edges_recursive():
+            if isinstance(parent_subgraph, SDFGState):
+                parent_sdfg = parent_subgraph.parent
+            elif isinstance(parent_subgraph, dstate.StateSubgraphView):
+                parent_sdfg = parent_subgraph.graph.parent
+            elif isinstance(parent_subgraph, SDFG):
+                # if there are any fancy things on the interstate edges we should probably throw an error
+                continue
+            else:
+                raise AutoDiffException("Unexpected subgraph structure")
+
+            if edge.data.data:
+                edge_type = parent_sdfg.arrays[edge.data.data].dtype
+                if edge_type not in [dace.float16, dace.float32, dace.float64]:
+                    raise AutoDiffException(
+                        f"Expected Subgraph to differentiate to only contain float edges, but data {edge.data}"
+                        f" on edge {edge} has type {edge_type}")
+
+        self._disambiguate_direction_dependent_views()
+
+        # recursively reverse the subgraph
+        self._reverse_subgraph(forward_subgraph)
+
+        self._applied = True
+
+        # in some cases (accessnode -> accessnode), the descriptors for the gradients of the function outputs are not
+        # added yet. Add them now
+
+        for given_grad in self.given_gradients:
+            if self.array_grad_name(
+                    given_grad.data) not in self.backward_sdfg.arrays:
+                self._add_gradient_data_descriptor(given_grad.data)
+
+        # prepare the output
+        required_grad_names = {
+            name.data: self.array_grad_name(name.data)
+            for name in self.required_gradients
+        }
+        given_grad_names = {
+            name.data: self.array_grad_name(name.data)
+            for name in self.given_gradients
+        }
+        result = BackwardResult(required_grad_names=required_grad_names,
+                                given_grad_names=given_grad_names)
+        return result, self.backward_grad_arrays, self.backward_input_arrays
+
+    def _find_subgraph_to_differentiate(self) -> dstate.StateSubgraphView:
+        """ Determine which nodes we need to reverse; this forms the subgraph we will differentiate:
+            we do a reverse bfs and a forward bfs, then take the intersection of nodes found
```

Comment:
please provide more explanation why (I thought that in reverse mode we just need to go from all the sinks backward)

Reply:
I've added a comment